### ISSUE TITLE
Clarify whether base64 encoding of license content is really optional

### ISSUE THREAD

**Issue by ben-sag** [CycloneDX/specification#382]
The specs (incl 1.4) are unclear about whether it is mandatory or optional to use base64 encoding for license text's "content". 

The fact the doc for "encoding" states it "must be one of"... "base64" makes me think it's mandatory to use base64, but the fact it also says "Specifies the **optional** encoding" makes me think maybe it's optional (e.g. if I don't specify an encoding parameter maybe it's treated as normal unencoded text). I looked at the examples but there is none that have content without encoding. 

Can you clarify, and in the next spec put something into the doc of the "content" field to make clear whether plain unencoded text is permitted? And maybe something in the examples to demonstrate unencoded text. 

(btw I'm very much hoping that use of base64 is not mandatory for this field - many licenses make it legally necessary to include the full license text and the SBOM is a very natural place to put that... but I doubt a law court would accept a base64 encoding of the text as meeting that requirement)

**Comment by mrutkows** [comment]
If you look at the spec./schema (i.e., specifically, https://cyclonedx.org/docs/1.5/json/#metadata_component_licenses_oneOf_i0_items_license_text_content) the `"content"` field (**the ONLY required field in the object**) is of type `string` with the description "The attachment data.  Proactive controls such as input validation and sanitization should be employed to prevent misuse of attachment text." which means it can be any  "text" data including (sanitized if possible) plain text.  

If the `"content"` is `"encoded"` then the ONLY value allowed for encoding is `"base64`" which means the `"content`" would need to be decoded prior to taking any further action using the "contentType". to interpret the decoded "text".  The stated default value for "contentType" is `"text/plain"`.

**Comment by ben-sag** [comment]
Thanks that's very helpful! I'm really glad non-encoded is an option. 

I did quite carefully read all the spec (including the part you quoted) yesterday - multiple times :)  - and couldn't convince myself 100% which way to interpret it (I found reasons for both possible interpretations :) ), so a brief remark in the next spec to make it explicit would be super helpful to avoid possible confusion. Perhaps add an example where the encode= is not specified. And in the "encoding" parameter you could state that the behaviour if the encoding parameter is not specified i.e. that the text is not encoded . (I realise that might sound too obvious to someone who already knows the intention of the spec, but the text stating it must be one of "base64" sent me off on a different interpretation. (I was also looking at this thread https://github.com/CycloneDX/cyclonedx-core-java/issues/100 where it looks like someone else concluded base64 was mandatory)

**Comment by ben-sag** [comment]
We're having a debate about this on https://github.com/CycloneDX/cyclonedx-core-java/issues/100 where it seems the main Java library for CycloneDX has taken a different interpretation of the encoding question.

**Comment by jkowalleck** [comment]
re https://github.com/CycloneDX/specification/issues/382#issuecomment-1954163503
> But I hope you would also consider supporting human-readable/non-base64 encoding in future - the spec allows for other legal texts such as copyrights to be human-readable, and the lack of supporting this for custom license texts makes the format unusable as a machine-readable way to comply with legal requirements.

@ben-sag ,
what do you mean? you can base64 any data, like a PDF `contentType=application/pdf`, or a plain text `contentType=text/plain`.
This means, you clearly can transport an arbitrary copyright test via CycloneDX, like so: 
```json
{"content": "KGMpIDIwMjQgSmFuZSBEb2Uu"}
```

**Comment by ben-sag** [comment]
Yes you can include any text, but it's not human readable.... so although you're 99% of the way to having a really useful file format that can both list the components _and_ fulfil the legal obligations (e.g. to publish license texts), the fact it's mandated to be in a non-human readable format (base64) stops the second use case. It's fine to allow base64 but forcing it as the only option is a surprising and imho unnecessary limitation. (Especially since you don't do that for other extracted text like copyrights but only license text).

**Comment by jkowalleck** [comment]
> Yes you can include any text, but it's not human readable....

let me ask it this way: would you give a JSON file to your lawyer? and would they open it as a pure text file and read the raw text? **literally** give a lawyer this example https://raw.githubusercontent.com/CycloneDX/bom-examples/master/SBOM/juice-shop/via_npm/bare/bom.1.4.json

> the fact it's mandated to be in a non-human readable format (base64) stops the second use case.

This is during transport. Nothing prevents you from base64-decoding the text, before handing it to some person.

> [...] don't do that for other extracted text like copyrights but only license text).

The field you are referring to is of type "attachment", not "text".
An "attachment" may be not only a license text; this could also be cryptographic assets, licensing information in a proprietary data format, some binary blob as an evidence for automation-based concluded data, ...


Please understand, that JSON and XML might be human-readable, but are not intended to be consumed by humans. They are data interchange formats for the digital age. 
These data are ingested by computer applications, like the web browser that you are currently using, which, in fact, received some data from github web servers, which was probably in no human-readable form during transport, and still, your web browser is able to display some text here ;-)

**Comment by jkowalleck** [comment]
@ben-sag I checked again and I was wrong.
As of CycloneDX 1.5, the encoding is indeed optional now, no default value anymore. 
Meaning this would be a valid attachment:
```json
{"content": "(c) 2024 Jan Doe"}
```

**Comment by sschuberth** [comment]
> As of CycloneDX 1.5, the encoding is indeed optional now, no default value anymore.

Hmm. Does that mean that depending on the specification / schema level chosen, a user of the [Java library](https://github.com/CycloneDX/cyclonedx-core-java) needs to manually encode e.g. license texts?

**Comment by stevespringett** [comment]
IMO, adding a method to the Java library that will encode/decode the license text would be a good feature to add.

**Comment by sschuberth** [comment]
> adding a method to the Java library that will encode/decode the license text

IMO, just adding such a function is not of much value, the Java code for that already is a one-liner. Instead, the library should automatically Base64-encode whatever is required to be Base64-encoded.

**Comment by ben-sag** [comment]
> > As of CycloneDX 1.5, the encoding is indeed optional now, no default value anymore.
> 
> Hmm. Does that mean that depending on the specification / schema level chosen, a user of the [Java library](https://github.com/CycloneDX/cyclonedx-core-java) needs to manually encode e.g. license texts?

I couldn't see any change to the spec - from v1.2 right through to v1.5 the encoding field is marked as "optional". 

So the Java library should not be doing any automatic transformations but just passing through the text unchanged. 

I do think given the confusion on this thread there's a very strong case now for a small change making it clear/explicit in the spec whether non-base64 content is permitted.

**Comment by mrutkows** [comment]
I believe (and asking as well) what Jan is saying, being the gatekeeper for v1.6, is that v1.6 it is no longer "optional"???  

However, the current JSON schema still indicates "[text](https://github.com/CycloneDX/specification/blob/472d4c2043a7c35854e11baa4488ba35156fc537/schema/bom-1.6.schema.json#L1206)" is of type "[attachment](https://github.com/CycloneDX/specification/blob/472d4c2043a7c35854e11baa4488ba35156fc537/schema/bom-1.6.schema.json#L1102)" which is still described as follows:

```json
   "attachment": {
      "type": "object",
      "title": "Attachment",
      "description": "Specifies the metadata and content for an attachment.",
      "required": [
        "content"
      ],
      ...
      "properties": {
        "contentType": {
          "type": "string",
          "title": "Content-Type",
          "description": "Specifies the content type of the text. Defaults to text/plain if not specified.",
          "default": "text/plain"
        },
        "encoding": {
          "type": "string",
          "title": "Encoding",
          "description": "Specifies the optional encoding the text is represented in.",
          "enum": [
            "base64"
          ],
          "meta:enum": {
            "base64": "Base64 is a binary-to-text encoding scheme that represents binary data in an ASCII string."
          }
        },
        "content": {
          "type": "string",
          "title": "Attachment Text",
          "description": "The attachment data. Proactive controls such as input validation and sanitization should be employed to prevent misuse of attachment text."
        }
      }
    },
```

which still as the only "required" field being "content" (i.e., "encoding" is not required) ... and the description of "encoding" still has "optional". 

Perhaps there is an outstanding (not merged) pull request (PR) that has this projected change?

**Comment by jkowalleck** [comment]
TLDR; 
Event with CDX 1.6, attachment encoding is not required, it is optional -- https://github.com/CycloneDX/specification/blob/2873c49989725d08faa6b2d14c332bc9da174348/schema/bom-1.6.schema.json#L1128-L1130
Therefore, omitting it means no encoding is applied.

----

> I believe (and asking as well) what Jan is saying, being the gatekeeper for v1.6, is that v1.6 it is no longer "optional"???

I was referring to a state when we had a schema issue during CycloneDX 1.3 JSON version only, when there was a default value for the "encoding" -- which was an unintended thing and was fixed in the CycloneDX JSON schema a while ago. This affected only JSON documents that were validated by the schema. 
This memory caused my confusion. Maybe by the fact that it drove me mad for some days, back then ... 😁 

The actual CycloneDX specification never had a default value for the attachment's "encodinng" field.
